atomics/include/desul/atomics/Generic.hpp

```diff
@@ -136,169 +136,88 @@ DESUL_INLINE_FUNCTION bool atomic_is_lock_free() noexcept {
       ;
 }
 
-template <class Oper, typename T, class MemoryOrder, class MemoryScope>
-DESUL_INLINE_FUNCTION T
-atomic_fetch_oper(const Oper& op,
-                  T* const dest,
-                  typename std::enable_if<sizeof(T) == 4, const T>::type val,
-                  MemoryOrder order,
-                  MemoryScope scope) {
-  union U {
-    int32_t i;
-    T t;
-    DESUL_INLINE_FUNCTION U() {}
-  } oldval, assume, newval;
-
-  oldval.t = *dest;
-
-  do {
-    assume.i = oldval.i;
-    newval.t = op.apply(assume.t, val);
-    oldval.i = desul::atomic_compare_exchange(
-        (int32_t*)dest, assume.i, newval.i, order, scope);
-  } while (assume.i != oldval.i);
-
-  return oldval.t;
-}
-
-template <class Oper, typename T, class MemoryOrder, class MemoryScope>
-DESUL_INLINE_FUNCTION T
-atomic_oper_fetch(const Oper& op,
-                  T* const dest,
-                  typename std::enable_if<sizeof(T) == 4, const T>::type val,
-                  MemoryOrder order,
-                  MemoryScope scope) {
-  union U {
-    int32_t i;
-    T t;
-    DESUL_INLINE_FUNCTION U() {}
-  } oldval, assume, newval;
-
-  oldval.t = *dest;
-
-  do {
-    assume.i = oldval.i;
-    newval.t = op.apply(assume.t, val);
-    oldval.i = desul::atomic_compare_exchange(
-        (int32_t*)dest, assume.i, newval.i, order, scope);
-  } while (assume.i != oldval.i);
-
-  return newval.t;
-}
-
-template <class Oper, typename T, class MemoryOrder, class MemoryScope>
-DESUL_INLINE_FUNCTION T
-atomic_fetch_oper(const Oper& op,
-                  T* const dest,
-                  typename std::enable_if<sizeof(T) == 8, const T>::type val,
-                  MemoryOrder order,
-                  MemoryScope scope) {
-  union U {
-    int64_t i;
-    T t;
-    DESUL_INLINE_FUNCTION U() {}
-  } oldval, assume, newval;
+template<std::size_t N>
+struct atomic_compare_exchange_type;
```

Comment:
```suggestion
template<std::size_t Bytes>
struct fixed_width_integer;
template<std::size_t Bytes>
using fixed_width_integer_t = typename fixed_width_integer<Bytes>::type;
```

Reply:
not sure that is good idea. I mean I have my dummy 16 byte thing which clearly isn't an integer ...

Reply:
`fixed_width_data` then?